src/platforms/esp32/components/avm_builtins/storage_nif.c

```diff
@@ -0,0 +1,304 @@
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <driver/sdmmc_host.h>
+#include <driver/sdspi_host.h>
+#include <esp_vfs_fat.h>
+
+#include <trace.h>
+
+#include "spi_driver.h"
+
+// TODO: allow ro option
+enum mount_type
+{
+    FATSPIFlash,
+    FATSDSPI,
+    FATSDMMC
+};
+
+struct MountedFS
+{
+    struct MiscEntry misc_entry;
+    char *base_path;
+    enum mount_type mount_type;
+    union
+    {
+        sdmmc_card_t *card;
+        wl_handle_t wl;
+    } handle;
+};
+
+static term make_esp_error_tuple(esp_err_t err, Context *ctx)
+{
+    if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    term result = term_alloc_tuple(2, &ctx->heap);
+    term_put_tuple_element(result, 0, ERROR_ATOM);
+    term_put_tuple_element(result, 1, esp_err_to_term(ctx->global, err));
+    return result;
+}
+
+static void opts_to_fatfs_mount_config(term opts_term, esp_vfs_fat_mount_config_t *mount_config)
+{
+    mount_config->format_if_mount_failed = true;
+    mount_config->max_files = 8;
+    mount_config->allocation_unit_size = 512;
+    // TODO: make it configurable: disk_status_check_enable = false
+}
+
+static term nif_esp_mount(Context *ctx, int argc, term argv[])
+{
+    term source_term = argv[0];
+    term target_term = argv[1];
+    term filesystem_type_term = argv[2];
+    term opts_term = argv[3];
+
+    int str_ok;
+    char *source = interop_term_to_string(source_term, &str_ok);
+    if (!str_ok) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    char *target = interop_term_to_string(target_term, &str_ok);
+    if (!str_ok || strlen(target) > 8) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    term fat_term
+        = globalcontext_existing_term_from_atom_string(ctx->global, ATOM_STR("\x3", "fat"));
+    if (term_is_invalid_term(fat_term) || filesystem_type_term != fat_term) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    if (!term_is_list(opts_term) && !term_is_map(opts_term)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    fprintf(stderr, "s: %s, t: %s\n", source, target);
+
+    struct ESP32PlatformData *platform = ctx->global->platform_data;
+
+    esp_vfs_fat_mount_config_t mount_config = {};
+    opts_to_fatfs_mount_config(opts_term, &mount_config);
+
+    esp_err_t ret = -1;
+
+    const char *part_by_name_prefix = "/dev/partition/by-name/";
+    int part_by_name_len = strlen(part_by_name_prefix);
+    if (!strncmp(part_by_name_prefix, source, part_by_name_len)) {
+        fprintf(stderr, "mount spi flash!!");
+        mount_config.allocation_unit_size = CONFIG_WL_SECTOR_SIZE;
+
+        struct MountedFS *mount = malloc(sizeof(struct MountedFS));
+        mount->misc_entry.entry_type = (uintptr_t) nif_esp_mount;
+        mount->base_path = strdup(target);
+        mount->mount_type = FATSPIFlash;
+        synclist_append(&platform->misc_entries, &mount->misc_entry.list_head);
+
+#if ESP_IDF_VERSION_MAJOR >= 5
+        ret = esp_vfs_fat_spiflash_mount(
```

Comment:
warning: 'esp_vfs_fat_spiflash_mount' is deprecated: esp_vfs_fat_spiflash_mount is deprecated, please use esp_vfs_fat_spiflash_mount_rw_wl instead [-Wdeprecated-declarations]